### PR TITLE
better code in generalized_kummer

### PR DESCRIPTION
**Pull request by fchapoton**
more readable, and should be also faster

### CODE REVIEW

File: diamond.py [hunk 1]
```diff
@@ -1951,11 +1951,20 @@ def generalised_kummer(n):
...
 
-        return HodgeDiamond.R(sum([gcd(list(a.keys()))**4 * (x*y)**(n - sum(a.values())) * prod([sum([prod([1 / HodgeDiamond.R((j**b[j]) * factorial(b[j])) * ((1-x**j) * (1-y**j))**(2*b[j]) for j in b.keys()]) for b in [b.to_exp_dict() for b in Partitions(a[i])]]) for i in a.keys()]) for a in [a.to_exp_dict() for a in Partitions(n)]])(-x, -y))
+    def my_Partitions(k):
+        for pi in Partitions(k):
+            yield pi.to_exp_dict()
+
+    def product(n):
+        hd = sum(gcd(a)**4 * (x*y)**(n - sum(a.values())) *
+                 prod([sum([prod([~((j**bj) * factorial(bj)) *
+                                  ((1 - x**j) * (1 - y**j))**(2*bj) for j, bj in b.items()])
```
Comment (pbelmans) [inline diamond.py:1964]: Why not write `b.to_exp_dict().items()` on this line, and avoid `my_Partitions(k)` altogether?

Comment (pbelmans) [reply diamond.py:1964]: Fair enough, on the next line it would be `a.to_exp_dict().values()`.

Comment (fchapoton) [reply diamond.py:1964]: the use of `my_Permutations` is a simple way to run through partitions without to ever create a list of partitions

Comment (pbelmans) [reply diamond.py:1964]: But replacing it with a call to Partitions and performing `to_exp_dict()` inline would also not create the full list of partitions, or am I mistaken?

Comment (fchapoton) [reply diamond.py:1964]: one could do what you suggest for b, but not for a, because a is called three times

**Comment by pbelmans** [comment]
I have now implemented your suggestion, combined with my suggestion to avoid `my_Partitions`, in a separate commit. Thanks!

**Comment by fchapoton** [comment]
ok, but this
```
for a in (ap.to_exp_dict(),))
```
is rather ugly..

**Comment by pbelmans** [comment]
It's my way of protesting the lack of a `let` keyword in Python, and I find it cleaner than the helper function.

Would you find `for a in [ap.to_exp_dict()]` cleaner?

**Comment by fchapoton** [comment]
there is now the walrus operator `:=` but I have never used that

https://realpython.com/python-walrus-operator/

**Comment by pbelmans** [comment]
Thanks to your suggestion, you have now indirectly used a walrus operator.

**Pull request closed without merging**